pyomo/contrib/doe/fim_doe.py

```diff
@@ -0,0 +1,2582 @@
+
+        # sipopt only uses backward difference scheme
+        # store measurements for scenarios
+        all_perturb_measure = []
+        all_base_measure = []
+        # store jacobian info
+        jac={}
+
+        # if measurements are provided
+        if read_output is not None:
+            with open(read_output, 'rb') as f:
+                output_record = pickle.load(f)
+                f.close()
+            jac = self.__finite_calculation(output_record, scena_gen)
+
+        else:
+            # time building time and solving time store list
+            time_allbuild = []
+            time_allsolve = []
+            # loop over parameters
+            for pa in range(len(self.param_name)):
+                perturb_mea = []
+                base_mea = []
+
+                # create model
+                time0_build = time.time()
+                mod = self.create_model(scenario_all, self.args)
+                time1_build = time.time()
+                time_allbuild.append(time1_build - time0_build)
+
+                # discretize if needed
+                if self.discretize_model is not None:
+                    mod = self.discretize_model(mod)
+
+                # For sIPOPT, fix model DOF
+                if self.mode =='sequential_sipopt':
+                    mod = self.__fix_design(mod, self.design_values, fix_opt=True)
+
+                # extract (discretized) time
+                time_set = []
+                for t in mod.t:
+                    time_set.append(value(t))
+
+                # add sIPOPT perturbation parameters
+                mod = self.__add_parameter(mod, perturb=pa)
+
+                # solve the square problem with the original parameters for k_aug mode, since k_aug does not calculate these
+                if self.mode == 'sequential_kaug':
+                    self.__solve_doe(mod, fix=True)
+
+                # parameter name lists for sipopt
+                list_original = []
+                list_perturb = []
+                for ele in self.param_name:
+                    list_original.append(getattr(mod, ele)[0])
+                for elem in self.perturb_names:
+                    list_perturb.append(getattr(mod, elem)[0])
+
+                # solve model
+                if self.mode =='sequential_sipopt':
+                    time0_solve = time.time()
+                    m_sipopt = sensitivity_calculation('sipopt', mod, list_original, list_perturb, tee=self.tee_opt, solver_options='ma57')
+                else:
+                    time0_solve = time.time()
+                    m_sipopt = sensitivity_calculation('k_aug', mod, list_original, list_perturb, tee=self.tee_opt, solver_options='ma57')
+
+                time1_solve = time.time()
+                time_allsolve.append(time1_solve - time0_solve)
+
+                # extract sipopt result
+                for j in self.flatten_measure_name:
+                    # check if this variable needs split name
```

Comment:
This does not need t be done in this PR, but we should add a note to revisit this in a future PR.  First, I think we should get away from storing "meaningful" information in the variable names (e.h., name and indixes, joined by a "magic string").  We should instead create simple names and then store an ancillary data structure that relates the simple name to the original name / index values

Second, we can probably get away from the two main indexed/unindexed branches here: Pyomo Scalar objects can be treated as indexed components (indexed by None)